Introduce BillingMode (1/4)

geropl opened the pull request:
## Description
This PR introduces `BiIllingMode`, which at it's core is an enum which tells you how a given user (and their usage) is billed. We use this abstraction to guide the (gradual) rollout of Usage-based Pricing, and later to phase out Chargebee.

This changeset does _not_ change any behavior, but rather adds the interface, methods to derive the current state for a given `User`/`Team` and tests for said implementation. It's 1 of 4 in a stack.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Context: https://github.com/gitpod-io/gitpod/issues/11402

## How to test
 - (cd components/server && leeway build .:dbtest)

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

## Werft options:
<!--
Optional annotations to add to the werft job.

* with-preview - whether to create a preview environment for this PR
-->
- [x] /werft with-preview
- [x] /werft with-payment

geropl commented:
/werft run

geropl commented:
/werft run

:crossed_fingers: test timeouts...

jankeromnes reviewed (approved):
Many thanks @geropl for powering through the first implementation and then the cleaner split! 💯

The code looks good to me (modulo some minor comments, please take a look and either address or dismiss), and the tests really build a ton of confidence here. Another thank you for always diligently adding lots of tests alongside your PRs 💯 I think these play a major role in Gitpod not falling apart as complexity grows. 💪

Would have been nice to be able to test this manually as well (e.g. with preview & payment & chargebee enabled, and double-checking the billing mode in various scenarii) but I guess we can do that in follow-up PRs too, so not a blocker.

/hold

jankeromnes commented:
One follow-up remark: As a sanity check, I tried to make your tests fail on purpose, by making a random change in the expectations: https://github.com/gitpod-io/gitpod/commit/89cde6114906f2d449d2955fd7e9106fd96b2880

However, it seems that my build went through without complaining. 😬 How can we be certain that the tests are actually executed?

geropl commented on components/gitpod-protocol/src/billing-mode.ts:
```diff
@@ -0,0 +1,67 @@
+    planIds: string[];
+
+    /** Whether or not the user has an active personal plan */
+    hasPersonalPlan?: boolean;
```
We identified this during our discussion on Friday: `mode` tells you how your sessions is going to be billed. But it's still unclear, whether or not we should show you the "Upgrade Stripe UI" on your personal billing page, for instance.
Example:
 - User has Chargebee Team seat, which "eats" all you bills => `mode: "chargebee"`
 - User has _no_ personal subscription: still able to sign up with Stripe (even though your invoice will stay empty, and  `mode: "chargebee"`)

geropl replied:
Addressed [here](https://github.com/gitpod-io/gitpod/pull/11812#issuecomment-1205029906).

geropl commented on components/gitpod-protocol/src/team-subscription-protocol.ts:
```diff
@@ -55,6 +55,9 @@ export namespace TeamSubscription2 {
     export const isActive = (ts2: TeamSubscription2, date: string): boolean => {
         return ts2.startDate <= date && (ts2.endDate === undefined || date < ts2.endDate);
     };
+    export function isCancelled(s: TeamSubscription2, date: string): boolean {
+        return (!!s.cancellationDate && s.cancellationDate < date) || (!!s.endDate && s.endDate < date); // This edge case is meant to handle bad data: If for whatever reason cancellationDate has not been set: treat endDate as such
+    }
```
> Duplicate from accounting-protocol?

Yes. Deemed it easier/more straight forward then abstracting over the field names/types.

> happy to merge as is then burn it all with fire when we finally deprecate Chargebee support

Absolutely. :fire: :innocent:

geropl commented on components/server/ee/src/billing/billing-mode.spec.db.ts:
```diff
@@ -0,0 +1,548 @@
+            },
+            // user: transition chargebee -> UBB
+            {
+                name: "user: chargbee paid personal (cancelled) + team seat",
```
- cancellationDate is the date when we click "cancel"
- endDate is when the subscription is no longer effective (currently: end of billing period)

Instead of setting endDate to "end of billing period", I think should prorate and set it to "cancellationDate + 24h", to give room for a safe transition period from Chargebee to Stripe (as discussed [here internally](https://www.notion.so/gitpod/Usage-Based-Pricing-Checkin-1185a6c70c1f4a93887b72cf5d225509#e196401e536a440aa91a804facad3280)).

:cloud: I'm just realizing that the current implementation here does not support that use-case very well, though. :cry: 
If you are still on a team seat, and deliberately want to upgrade to "personal usage-based", then you also want to ripe the benefits of that. @jankeromnes Would be awesome if we can find 10mins of sync time on this and adjust.

geropl replied:
Addressed [here](https://github.com/gitpod-io/gitpod/pull/11812#issuecomment-1205029906).

geropl commented on components/gitpod-protocol/src/billing-mode.ts:
```diff
@@ -0,0 +1,67 @@
+    | UBBFreeTier
+    | UBBPaidTier;
+
+export namespace BillingMode {
+    export const NONE: BillingMode = {
+        mode: "none",
+    };
+}
```
What do you mean? I'd like to use `BillingMode.NONE` somewhere... I guess I'm missing sth?

jankeromnes replied:
I guess I was just wondering why the `None` mode is defined twice (once here, and again just below -- seems slightly risky e.g. if we tweak it later but forget to update the other copy)

geropl replied:
Good point. Addressed by changing type to `None` :upside_down_face:

AlexTugarev commented:
/werft run

geropl commented:
@AlexTugarev After discussing with @jankeromnes I'm actively working on further simplyfing this. sorry for not communicating earlier on the PR :confused:

geropl commented:
@jankeromnes I came up with a new approach which addresses all of your comments [here](https://github.com/gitpod-io/gitpod/pull/11812#pullrequestreview-1059837319) (complexity), [here](https://github.com/gitpod-io/gitpod/pull/11812#discussion_r936444652) (cleared modifiers: cmp. `canUpgradeToUBB`) and [here](https://github.com/gitpod-io/gitpod/pull/11812#discussion_r936453521).

The core change is that UBB now is "gready": If you have at least one UBB team Seat, or a personal UBB subscription, that "overrides" your Chargebee one. This way we can ensure you receive the benefits of UBB (workspace classes) immediately.
In the end this makes the code far nicer as well. :lotus_position: :upside_down_face:

geropl commented on components/gitpod-protocol/src/billing-mode.ts:
```diff
@@ -0,0 +1,66 @@
+        mode: "none",
+    };
+
+    export function showChargebeeBilling(billingMode: BillingMode): boolean {
```
These functions are not used nor tested, yet, but helped me greatly when thinking about how we want to used BillingMode later on. Left them in for reference, and maybe they're actually helpful

jankeromnes commented on components/server/ee/src/billing/billing-mode.ts:
```diff
@@ -0,0 +1,174 @@
+
+        // 1. UBB enabled?
+        if (!isUsageBasedBillingEnabled) {
+            // No Chargebee subscription, and UBB is not enabled: chargebee (free)
```
Nit: This comment is inaccurate (we don't know that here are no chargebee subscriptions because we didn't check yet)

jankeromnes commented on components/server/ee/src/billing/billing-mode.ts:
```diff
@@ -0,0 +1,174 @@
+        // 2. Any Chargbee TeamSubscription2 (old Team Subscriptions are not relevant here, as they are not associated with a team)
+        const teamSubscription = await this.teamSubscription2Db.findForTeam(team.id, now);
+        if (teamSubscription && TeamSubscription2.isActive(teamSubscription, now)) {
+            if (isUsageBasedBillingEnabled && TeamSubscription2.isCancelled(teamSubscription, now)) {
```
Nit: I think you don't need to check `isUsageBasedBillingEnabled` here (we know it's `true`, because otherwise we would have already bailed out in the UBB check above)

jankeromnes commented:
> ## How to test
> 
> - (cd components/server && leeway build .:db-test)

Oops:

```
$ cd components/server

$ leeway build .:db-test
FATA[0000] package "components/server:db-test" does not exist
```

jankeromnes commented:
Thanks!

Cannot double-approve: Please fix the two conflicts and feel free to merge! 🛹 😁

geropl commented:
> Cannot double-approve: Please fix the two conflicts and feel free to merge! skateboard grin

Thank you, @jankeromnes!  :rocket: :pray:

geropl commented:
/unhold

Did rebase, and ran tests again :upside_down_face:
